csrc/preseg_passes/propagate_shardings.cpp

```diff
@@ -312,16 +316,33 @@ void propagateDIDTransform(
     TensorView* tv,
     const std::unordered_set<ParallelType>& selected_parallel_types,
     PropagateDirection direction) {
-  tv->setDeviceMesh(ref->getDeviceMesh());
-
   std::unordered_set<IterDomain*> device_ids;
   const std::unordered_map<IterDomain*, IterDomain*> ref2target =
       getRef2TargetMap(ref, tv, direction);
 
+  const auto& ref_mesh = ref->getDeviceMesh();
+
   for (IterDomain* device_id : ref->getLoopDomain()) {
     if (selected_parallel_types.count(device_id->getParallelType()) == 0) {
       continue;
     }
+
+    // If the target has a device mesh, the device id can only be propagated
+    // iff:
+    // 1. The device id parallel type is present in the device mesh.
+    // 2. The number of devices corresponding to the parallel type is same
+    // between the reference and target.
+    if (tv->hasDeviceMesh()) {
+      const auto& target_mesh = tv->getDeviceMesh();
+      const auto parallel_type = device_id->getParallelType();
+      if (!target_mesh.hasParallelType(parallel_type)) {
+        continue;
+      }
+      if (ref_mesh.size(parallel_type) != target_mesh.size(parallel_type)) {
```

Comment:
Is this check necessary? IIRC, we only check logically-mapped dimensions which ought to have the same size.

Reply:
This is meant to check the number of devices corresponding to the parallel type, aka the split factor.
For eg:
```
tv0 = makeContigTensor(2)
tv1 = makeContigTensor(2)
tv2 = tv0 + tv1

tv0 DeviceMesh(dx, 2*dy)
tv0->outer_split(0, dx)
tv0->axis(0)->parallelize(ParallelType::DIDx)

tv1 DeviceMesh(dx, dy)
tv1->outer_split(1, dy)
tv1->axis(1)->parallelize(ParallelType::DIDy)
```

When we propagate from `tv0->tv2`, `tv2` has the same device mesh as `tv0`. This check verifies that when propagating from `tv1`, the number of device, aka, split factor for DIDy is consistent with the existing device mesh.

Does that clarify the intent here?